tasks/fine_tuning_sbert/src/loops.py

```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
-                print("Time taken for fine-tuning:", "{:0>2}:{:0>2}:{:05.2f}".format(int(hours), int(minutes), seconds))
+    X_train, X_dev, y_train, y_dev = train_test_split(train_sents, train_labels, test_size=dev_perc,
+                                                      stratify=train_labels, random_state=100)
+
+    # Load data samples into batches
+    train_batch_size = 16
+    train_samples = build_data_samples(X_train, label2int, y_train)
+    dev_samples = build_data_samples(X_dev, label2int, y_dev)
+
+    # Train set config
+    model = EarlyStoppingSentenceTransformer(model_name)
+    train_dataset = SentencesDataset(train_samples, model=model)
+    train_dataloader = DataLoader(
+        train_dataset, shuffle=True, batch_size=train_batch_size)
+
+    # Dev set config
+    dev_dataset = SentencesDataset(dev_samples, model=model)
+    dev_dataloader = DataLoader(
+        dev_dataset, shuffle=True, batch_size=train_batch_size)
+
+    # Define the way the loss is computed
+    classifier = SoftmaxClassifier(model=model,
+                                   sentence_embedding_dimension=model.get_sentence_embedding_dimension(),
+                                   num_labels=len(label2int))
+    warmup_steps = math.ceil(
+        len(train_dataset) * max_num_epochs / train_batch_size * 0.1)  # 10% of train data for warm-up
+
+    set_seeds(seed)
+    model_deets = f"{train_params['eval_classifier']}_model={model_name}_test-perc={dev_perc}_seed={seed}"
+
+    # Train the model
+    start = time.time()
+    dev_evaluator = CustomLabelAccuracyEvaluator(dataloader=dev_dataloader, softmax_model=classifier,
+                                                 name='lae-dev', label_names=label_names,
+                                                 model_hyper_params={'model_name': model_name, 'dev_perc': dev_perc, 'seed': seed})
+
+    # this will write to the same project every time
+    run = wandb.init(notes=model_deets, project='WRI', tags=['baseline', 'training'],
+                     entity='ramanshsharma')
+
+    model.fit(train_objectives=[(train_dataloader, classifier)],
+              evaluator=dev_evaluator,
+              epochs=max_num_epochs,
+              evaluation_steps=1000,
+              warmup_steps=warmup_steps,
+              output_path=output_path,
+              model_deets=model_deets,
+              baseline=baseline,
+              patience=patience,
+              show_progress_bar=False
+              )
+
+    run.save()
+    run_name = run.id
+
+    torch.save(model, output_path+'/saved_model.pt')
+    wandb.save(output_path+'/saved_model.pt')
+
+    wandb.finish()
+
+    end = time.time()
+    hours, rem = divmod(end - start, 3600)
+    minutes, seconds = divmod(rem, 60)
+    print("Time taken for fine-tuning:",
+          "{:0>2}:{:0>2}:{:05.2f}".format(int(hours), int(minutes), seconds))
+
+    return run_name
+
+
+def make_dataset_public(train_sents_, train_labels_, label_names_):
```

Comment:
I'm not gonna lie, I get scared of global variables but in this case, well done - it seems like a good solution for now :)

Reply:
Yeah I kept thinking and found this to be a way to communicate the dataset between the notebook and the ```src``` files :D But for sure we can come up with a better solution for further experiments